x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authc/service/ServiceAccountService.java

```diff
@@ -0,0 +1,155 @@
+
+    public ServiceAccountService(ServiceAccountsCredentialStore serviceAccountsCredentialStore) {
+        this.serviceAccountsCredentialStore = serviceAccountsCredentialStore;
+    }
+
+    public static boolean isServiceAccount(Authentication authentication) {
+        return REALM_TYPE.equals(authentication.getAuthenticatedBy().getType());
+    }
+
+    // {@link org.elasticsearch.xpack.security.authc.TokenService#extractBearerTokenFromHeader extracted} from an HTTP authorization header.
+    /**
+     * Parses a token object from the content of a {@link ServiceAccountToken#asBearerString()} bearer string}.
+     * This bearer string would typically be
+     *
+     * <p>
+     * <strong>This method does not validate the credential, it simply parses it.</strong>
+     * There is no guarantee that the {@link ServiceAccountToken#getSecret() secret} is valid,
+     * or even that the {@link ServiceAccountToken#getAccountId() account} exists.
+     * </p>
+     * @param token A raw token string (if this is from an HTTP header, then the <code>"Bearer "</code> prefix must be removed before
+     *              calling this method.
+     * @return An unvalidated token object.
+     */
+    public static ServiceAccountToken tryParseToken(SecureString token) {
+        try {
+            if (token == null) {
+                return null;
+            }
+            return doParseToken(token);
+        } catch (IOException e) {
+            logger.debug("Cannot parse possible service account token", e);
+            return null;
+        }
+    }
+
+    public void authenticateWithToken(ServiceAccountToken token, ThreadContext threadContext, String nodeName,
+                                      ActionListener<Authentication> listener) {
+
+        if (ElasticServiceAccounts.NAMESPACE.equals(token.getAccountId().namespace()) == false) {
+            logger.debug("only [{}] service accounts are supported, but received [{}]",
+                ElasticServiceAccounts.NAMESPACE, token.getAccountId().asPrincipal());
+            listener.onResponse(null);
+            return;
+        }
+
+        final ServiceAccount account = ACCOUNTS.get(token.getAccountId().serviceName());
+        if (account == null) {
+            logger.debug("the [{}] service account does not exist", token.getAccountId().asPrincipal());
+            listener.onFailure(null);
+            return;
+        }
+
+        if (serviceAccountsCredentialStore.authenticate(token)) {
+            listener.onResponse(success(account, token, nodeName));
+        } else {
+            final ParameterizedMessage message = new ParameterizedMessage(
+                "failed to authenticate service account [{}] with token name []",
+                token.getAccountId().asPrincipal(),
+                token.getTokenName());
+            logger.debug(message);
+            throw new ElasticsearchSecurityException(message.getFormattedMessage());
```

Comment:
Despite what I say about throwing below, I think this should be `onFailure(new ElasticsearchSecurityException(..))` instead.

Reply:
I thought about that, but it seems we mostly use `onFailure` for system related issues, e.g. security index is not available etc. Also because I plan to refactor this later to return an updated `AuthenticationResult` which uses terminate instead of `onFailure` for credential mismatch. So overall I think throwing is ok here and all throwing will become `AuthenticationResult#terminate` in future. Does it make sense?

Reply:
> mostly use onFailure for system related issues

That's not intentional. `onFailure` is for when exceptions happen in async code.

So once `serviceAccountsCredentialStore.authenticate` becomes asynchronous, this will need to be `onFailure` because it's too late to throw an exception - control has passed from the original thread to the callback.

In general, it is preferable for code that takes a listener argument to signal all outcomes via `onFailure` or `onResponse` and never `throw` or `return`